### PR TITLE
feat: Update migrate_users script to also update team memberships.

### PR DESCRIPTION
**Pull request by feanil**
Some users are already in the destination openedx org but were not
already added to relevant teams or may have been added to new teams
since they were migrated.

For any users that are full members of the destination org, ensure that
they are members of all relevant teams.

### CODE REVIEW

File: migrate/migrate_users.py [hunk 1]
```diff
@@ -177,6 +189,20 @@ def migrate(
                     )
                 time.sleep(wait_seconds)
 
+    for index, username in enumerate(requested_users_in_org):
```
Comment (kdmccormick) [inline migrate/migrate_users.py:192]:
```suggestion
    for index, username in enumerate(requested_users_already_in_org):
```

File: migrate/migrate_users.py [hunk 2]
```diff
@@ -177,6 +189,20 @@ def migrate(
...
+        team_slugs = user_team_slugs[username]
+        click.echo(
+            f"({index:03d}/{len(requested_users_in_org)}) updating teams for user {username}; "
```
Comment (kdmccormick) [inline migrate/migrate_users.py:195]:
```suggestion
            f"({index:03d}/{len(requested_users_already_in_org)}) updating teams for user {username}; "
```

File: migrate/migrate_users.py [hunk 3]
```diff
@@ -89,10 +89,14 @@ def migrate(
...
 
     # Build a mapping from usernames to team slugs.
+    # Needed by the invitation endpoint
     user_teams: Dict[str, Set[int]] = defaultdict(set)
+    # Needed for the update user membership endpoint
+    user_team_slugs: Dict[str, Set[str]] = defaultdict(set)
     for team_slug, members in team_users.items():
         for username in members:
             user_teams[username].add(team_slugs_to_ids[team_slug])
+            user_teams[username].add(team_slug)
```
Comment (kdmccormick) [inline migrate/migrate_users.py:99]:
```suggestion
            user_team_slugs[username].add(team_slug)
```

**Review by kdmccormick — COMMENTED** [review]
Some typos, but the logic looks good.

File: migrate/migrate_users.py [hunk 4]
```diff
@@ -177,6 +189,20 @@ def migrate(
...
+            f"({index:03d}/{len(requested_users_already_in_org)}) updating teams for user {username}; "
+            f"{len(team_slugs)=}."
+        )
+        for team_slug in team_slugs:
+            api.teams.add_or_update_membership_for_user_in_org(
+                org=dest_org,
+                team_slug=team_slug,
+                username=username,
+                role="member",
+            )
```
Comment (nedbat) [inline migrate/migrate_users.py:204]: We need to test what happens if the user is already a member of the team.

**Review by kdmccormick — APPROVED** [review]

File: migrate/migrate_users.py [hunk 5]
```diff
@@ -188,20 +212,65 @@ def extract_user_names(user_list_file) -> List[str]:
     return empty_lines_removed
 
 
-def extract_merged_team_memberships(org_export_files: list) -> Dict[str, Set[str]]:
+@dataclass(frozen=True)
+class Team:
```
Comment (kdmccormick) [inline migrate/migrate_users.py:216]: I think `TeamMembership` would describe this entity more precisely.

File: migrate/migrate_users.py [hunk 6]
```diff
@@ -188,20 +212,62 @@ def extract_user_names(user_list_file) -> List[str]:
...
+                    # because the user is already on the team with maintainer level access
+                    # via another org.
+                users_to_teams[member].add(member_team)
+
+            for maintainer in team.get("maintainers", []):
+                if member_team in users_to_teams[maintainer]:
+                    users_to_teams[maintainer].remove(member_team)
+                    # because we're gonna elevate this user's access
+
+                users_to_teams[maintainer].add(maintainer_team)
```
Comment (kdmccormick) [inline migrate/migrate_users.py:268]: `members` and `maintainers` in the export are disjoint, so you could simplify this logic if you wanted to.

**Pull request merged**